crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -283,42 +278,71 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
-                            let ty = IntersectionBuilder::new(self.db)
-                                .add_negative(rhs_ty)
-                                .build();
-                            constraints.insert(symbol, ty);
-                        } else {
-                            // Non-singletons cannot be safely narrowed using `is not`
+                    match if is_positive { *op } else { op.negate() } {
+                        ast::CmpOp::IsNot => {
+                            if rhs_ty.is_singleton(self.db) {
+                                let ty = IntersectionBuilder::new(self.db)
+                                    .add_negative(rhs_ty)
+                                    .build();
+                                constraints.insert(symbol, ty);
+                            } else {
+                                // Non-singletons cannot be safely narrowed using `is not`
+                            }
                         }
-                    }
-                    ast::CmpOp::Is => {
-                        constraints.insert(symbol, rhs_ty);
-                    }
-                    ast::CmpOp::NotEq => {
-                        if rhs_ty.is_single_valued(self.db) {
-                            let ty = IntersectionBuilder::new(self.db)
-                                .add_negative(rhs_ty)
-                                .build();
-                            constraints.insert(symbol, ty);
+                        ast::CmpOp::Is => {
+                            constraints.insert(symbol, rhs_ty);
+                        }
+                        ast::CmpOp::NotEq => {
+                            if rhs_ty.is_single_valued(self.db) {
+                                let ty = IntersectionBuilder::new(self.db)
+                                    .add_negative(rhs_ty)
+                                    .build();
+                                constraints.insert(symbol, ty);
+                            }
+                        }
+                        _ => {
+                            // TODO other comparison types
                         }
                     }
-                    _ => {
-                        // TODO other comparison types
+                }
+                ast::Expr::Call(ast::ExprCall {
+                    range: _,
+                    func: callable,
+                    arguments,
+                }) => {
+                    let callable_ty =
+                        inference.expression_ty(callable.scoped_expression_id(self.db, scope));
+
+                    if callable_ty
+                        .into_class_literal()
+                        .map(|c| c.class.is_known(self.db, KnownClass::Type))
+                        .unwrap_or(false)
+                        && arguments.len() == 1
+                    {
+                        if let Some(id) = arguments.args[0].as_name_expr().map(|name| &name.id) {
```

Comment:
This panics with an index out of bounds if I make this modification to your tests:

```diff
--- a/crates/red_knot_python_semantic/resources/mdtest/narrow/type.md
+++ b/crates/red_knot_python_semantic/resources/mdtest/narrow/type.md
@@ -10,7 +10,7 @@ def get_a_or_b() -> A | B: ...
 
 x = get_a_or_b()
 
-if type(x) is A:
+if type(object=x) is A:
     reveal_type(x)  # revealed: A
 else:
     # It would be wrong to infer `B` here. The type
```

Which is an invalid call, but we shouldn't panic on it

Reply:
This is because you check `arguments.len() == 1` above, but that will be true if there's exactly one _keyword_ argument: https://github.com/astral-sh/ruff/blob/1f07880d5c410bcd5192a1bd20e8f7d59bc4aab4/crates/ruff_python_ast/src/nodes.rs#L3777-L3781